### PR TITLE
Fix Subscription Processing Workaround

### PR DESCRIPTION
**Pull request by dylanbronson**
Removing a workaround that we added for 1.0.9 that fixed an issue we were having that when an MQTT identity with only subscriptions (no telemetry) reauthenticated, the identity wouldn't reconnect to the cloud. (PR of previous problem: https://github.com/Azure/iotedge/pull/3201)

The new way of doing this is to listen to when a device gets connected to EdgeHub and process subscriptions when it does.
To prevent a re-triggering of subscription processing when EdgeHub connects to the cloud on behalf of the device (this situation occurs during the device connecting to edgeHub flow), we use a ConcurrentDictionary to keep track of if subscription processing has already been kicked off for a DeviceConnecting. We check this in the other methods. 
This effectively ignores the Cloud Connection event trigger.

### CODE REVIEW

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/SubscriptionProcessor.cs [hunk 1]
```diff
@@ -44,11 +45,34 @@ public SubscriptionProcessor(IConnectionManager connectionManager, IInvokeMethod
             Preconditions.CheckNotNull(deviceConnectivityManager, nameof(deviceConnectivityManager));
             this.invokeMethodHandler = Preconditions.CheckNotNull(invokeMethodHandler, nameof(invokeMethodHandler));
             this.pendingSubscriptions = new ConcurrentDictionary<string, ConcurrentQueue<(DeviceSubscription, bool)>>();
+            this.subscriptionsBeingProcessed = new ConcurrentDictionary<string, bool>();
```
Comment (varunpuranik) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/SubscriptionProcessor.cs:48]: Nit: Initialize inline

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/SubscriptionProcessor.cs [hunk 2]
```diff
@@ -44,11 +45,34 @@ public SubscriptionProcessor(IConnectionManager connectionManager, IInvokeMethod
...
+                Events.ProcessingSubscriptionsOnDeviceConnectedToEdgeHub(identity);
+                await this.ProcessExistingSubscriptions(identity.Id);
+            }
+            catch (Exception e)
+            {
+                Events.ErrorProcessingSubscriptions(e, identity);
+            }
+            finally
+            {
+                this.ConnectionManager.CloudConnectionEstablished += this.ClientCloudConnectionEstablished;
```
Comment (varunpuranik) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/SubscriptionProcessor.cs:71]: Why is this needed?

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/SubscriptionProcessor.cs [hunk 3]
```diff
@@ -44,11 +45,34 @@ public SubscriptionProcessor(IConnectionManager connectionManager, IInvokeMethod
...
             this.invokeMethodHandler = Preconditions.CheckNotNull(invokeMethodHandler, nameof(invokeMethodHandler));
             this.pendingSubscriptions = new ConcurrentDictionary<string, ConcurrentQueue<(DeviceSubscription, bool)>>();
+            this.subscriptionsBeingProcessed = new ConcurrentDictionary<string, bool>();
             this.processSubscriptionsBlock = new ActionBlock<string>(this.ProcessPendingSubscriptions);
+            connectionManager.DeviceConnected += this.DeviceConnected;
             deviceConnectivityManager.DeviceConnected += this.CloudConnectivityEstablished;
             connectionManager.CloudConnectionEstablished += this.ClientCloudConnectionEstablished;
         }
 
+        async void DeviceConnected(object sender, IIdentity identity)
```
Comment (varunpuranik) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/SubscriptionProcessor.cs:55]: Instead of this, can we use the same callback for both DeviceConnected and CloudConnectionEstablished and if an identity is being processed, then skip processing it again?

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/SubscriptionProcessor.cs [hunk 4]
```diff
@@ -142,11 +144,29 @@ async Task ProcessSubscriptionByIdentity(IIdentity identity)
...
+        void ClientConnectionToEdgeHubEstablished(object sender, IIdentity identity)
+        {
+            this.ClientConnectionEstablished(identity, false);
+        }
+
+        async void ClientConnectionEstablished(IIdentity identity, bool connectedToCloud)
+        {
+            if (connectedToCloud)
+            {
+                Events.ClientConnectedToCloudProcessingSubscriptions(identity);
```
Comment (varunpuranik) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/SubscriptionProcessor.cs:161]: Nit: Move the events to the parent and remove this method.

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/SubscriptionProcessor.cs [hunk 5]
```diff
@@ -142,11 +144,29 @@ async Task ProcessSubscriptionByIdentity(IIdentity identity)
...
+        }
+
+        void ClientConnectionToEdgeHubEstablished(object sender, IIdentity identity)
+        {
+            this.ClientConnectionEstablished(identity, false);
+        }
+
+        async void ClientConnectionEstablished(IIdentity identity, bool connectedToCloud)
+        {
+            if (connectedToCloud)
```
Comment (ancaantochi) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/SubscriptionProcessor.cs:159]: nit: connectedToCloud could be passed to the Events method so the if can be inside it

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/SubscriptionProcessor.cs [hunk 6]
```diff
@@ -157,16 +177,27 @@ async void ClientCloudConnectionEstablished(object sender, IIdentity identity)
...
                     {
-                        await this.ProcessSubscriptionWithRetry(id, cloudProxy, subscription.Key, subscription.Value);
-                    }
-                });
+                        foreach (KeyValuePair<DeviceSubscription, bool> subscription in s)
+                        {
+                            await this.ProcessSubscriptionWithRetry(id, cloudProxy, subscription.Key, subscription.Value);
+                        }
+                    });
+                this.subscriptionsBeingProcessed.TryRemove(id, out _);
```
Comment (ancaantochi) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/SubscriptionProcessor.cs:199]: should have try finally and remove in finally to make sure it gets removed after it was processed

Comment (varunpuranik) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/SubscriptionProcessor.cs:199]: +1
